File: Parallel-Java/src/parallel/containers/UnorderedMapParallel.java
package parallel.containers;

import org.javatuples.Pair;

import java.util.*;

public class UnorderedMapParallel {
    // T1
    public HashMap<Character, Character> getMapChar(HashMap<Character, Character> map){
        HashMap<Character, Character> m = new HashMap<>(Map.of('a', 'a',
                'b', 'b', 'c', 'c'));
        map = new HashMap<>(m);
        return m;
    }

    // T2
    public void clearMapChar(HashMap<Character, Character> map, int size){
        if(map.isEmpty()) return;
        if(map.size() <= size) map.clear();
    }

    // T3
    public void containsThreeChar(HashMap<Character, Character> map){
        if(map.containsKey('a') && map.containsKey('b') && map.containsKey('c')){
            map.put('d','d');
        }
    }

    // T4
    public void containsDeleteChar(HashMap<Character, Character> map, Character element){
        if(map.containsKey(element)){
            map.remove(element);
        }
    }

    public void containsDeleteListChar(HashMap<Character, Character> map, List<Character> deleteList){
        for(Character element : deleteList){
            if(map.containsKey(element)){
                map.remove(element);
            }
        }
    }

    // T5
    public void tryAddSizeChar(HashMap<Character, Character> map, int size, Character c){
        if(map.size() >= size) return;
        map.putIfAbsent(c, c);
    }

    public void tryAddListChar(HashMap<Character, Character> map, List<Character> list){
        for(Character element : list){
            map.putIfAbsent(element, element);
        }
    }

    // T6
    public List<Character> getValuesListCharWithDefault(HashMap<Character, Character> map, List<Character> keys, Character default_){
        List<Character> values = new ArrayList<>();
        for(Character key : keys){
            Character value = map.getOrDefault(key, default_);
            values.add(value);
        }
        return values;
    }

    // T7
    public List<Optional<Character>> getValuesListCharWithNull(HashMap<Character, Character> map, List<Character> keys){
        List<Optional<Character>> values = new ArrayList<>();
        for(Character key : keys){
            Optional<Character> value = Optional.ofNullable(map.get(key));
            values.add(value);
        }
        return values;
    }

    // T8
    public HashMap<Character, Character> composeMapChar(List<Character> keys, List<Character> values){
        HashMap<Character, Character> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<Character, String> composeMapCharString(List<Character> keys, List<String> values){
        HashMap<Character, String> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<Character, Character> composeMapsChar(HashMap<Character, Character> map1, HashMap<Character, Character> map2){
        HashMap<Character, Character> map = new HashMap<>();
        for(Map.Entry<Character, Character> ent : map1.entrySet()){
            map.put(ent.getKey(), ent.getValue());
        }

        for(Map.Entry<Character, Character> ent : map2.entrySet()){
            map.putIfAbsent(ent.getKey(), ent.getValue());
        }
        return map;
    }

    // Convert Char to String

    // T1
    public HashMap<String, String> getMapString(HashMap<String, String> map){
        HashMap<String, String> m = new HashMap<>(Map.of("aa", "aa",
                "bb", "bb", "cc", "cc"));
        map = new HashMap<>(m);
        return m;
    }

    // T2
    public void clearMapString(HashMap<String, String> map, int size){
        if(map.isEmpty()) return;
        if(map.size() <= size) map.clear();
    }

    // T3
    public void containsThreeString(HashMap<String, String> map){
        if(map.containsKey("aa") && map.containsKey("bb") && map.containsKey("cc")){
            map.put("dd","dd");
        }
    }

    // T4
    public void containsDeleteString(HashMap<String, String> map, String element){
        if(map.containsKey(element)){
            map.remove(element);
        }
    }

    public void containsDeleteListString(HashMap<String, String> map, List<String> deleteList){
        for(String element : deleteList){
            if(map.containsKey(element)){
                map.remove(element);
            }
        }
    }

    // T5
    public void tryAddSizeString(HashMap<String, String> map, int size, String c){
        if(map.size() >= size) return;
        map.putIfAbsent(c, c);
    }

    public void tryAddListString(HashMap<String, String> map, List<String> list){
        for(String element : list){
            map.putIfAbsent(element, element);
        }
    }

    // T6
    public List<String> getValuesListStringWithDefault(HashMap<String, String> map, List<String> keys, String default_){
        List<String> values = new ArrayList<>();
        for(String key : keys){
            String value = map.getOrDefault(key, default_);
            values.add(value);
        }
        return values;
    }

    // T7
    public List<Optional<String>> getValuesListStringWithNull(HashMap<String, String> map, List<String> keys){
        List<Optional<String>> values = new ArrayList<>();
        for(String key : keys){
            Optional<String> value = Optional.ofNullable(map.get(key));
            values.add(value);
        }
        return values;
    }

    // T8
    public HashMap<String, String> composeMapString(List<String> keys, List<String> values){
        HashMap<String, String> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<String, Character> composeMapStringChar(List<String> keys, List<Character> values){
        HashMap<String, Character> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<String, String> composeMapsString(HashMap<String, String> map1, HashMap<String, String> map2){
        HashMap<String, String> map = new HashMap<>();
        for(Map.Entry<String, String> ent : map1.entrySet()){
            map.put(ent.getKey(), ent.getValue());
        }

        for(Map.Entry<String, String> ent : map2.entrySet()){
            map.putIfAbsent(ent.getKey(), ent.getValue());
        }
        return map;
    }

    // Convert Char to Int
    // T1
    public HashMap<Integer, Integer> getMapInt(HashMap<Integer, Integer> map){
        HashMap<Integer, Integer> m = new HashMap<>(Map.of(1, 1,
                2, 2, 3, 3));
        map = new HashMap<>(m);
        return m;
    }

    // T2
    public void clearMapInt(HashMap<Integer, Integer> map, int size){
        if(map.isEmpty()) return;
        if(map.size() <= size) map.clear();
    }

    // T3
    public void containsThreeInt(HashMap<Integer, Integer> map){
        if(map.containsKey(1) && map.containsKey(2) && map.containsKey(3)){
            map.put(4,4);
        }
    }

    // T4
    public void containsDeleteInt(HashMap<Integer, Integer> map, Integer element){
        if(map.containsKey(element)){
            map.remove(element);
        }
    }

    public void containsDeleteListInt(HashMap<Integer, Integer> map, List<Integer> deleteList){
        for(Integer element : deleteList){
            if(map.containsKey(element)){
                map.remove(element);
            }
        }
    }

    // T5
    public void tryAddSizeInt(HashMap<Integer, Integer> map, int size, Integer c){
        if(map.size() >= size) return;
        map.putIfAbsent(c, c);
    }

    public void tryAddListInt(HashMap<Integer, Integer> map, List<Integer> list){
        for(Integer element : list){
            map.putIfAbsent(element, element);
        }
    }

    // T6
    public List<Integer> getValuesListIntWithDefault(HashMap<Integer, Integer> map, List<Integer> keys, Integer default_){
        List<Integer> values = new ArrayList<>();
        for(Integer key : keys){
            Integer value = map.getOrDefault(key, default_);
            values.add(value);
        }
        return values;
    }

    // T7
    public List<Optional<Integer>> getValuesListIntWithNull(HashMap<Integer, Integer> map, List<Integer> keys){
        List<Optional<Integer>> values = new ArrayList<>();
        for(Integer key : keys){
            Optional<Integer> value = Optional.ofNullable(map.get(key));
            values.add(value);
        }
        return values;
    }

    // T8
    public HashMap<Integer, Integer> composeMapInt(List<Integer> keys, List<Integer> values){
        HashMap<Integer, Integer> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<Integer, String> composeMapIntString(List<Integer> keys, List<String> values){
        HashMap<Integer, String> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<Integer, Integer> composeMapsInt(HashMap<Integer, Integer> map1, HashMap<Integer, Integer> map2){
        HashMap<Integer, Integer> map = new HashMap<>();
        for(Map.Entry<Integer, Integer> ent : map1.entrySet()){
            map.put(ent.getKey(), ent.getValue());
        }

        for(Map.Entry<Integer, Integer> ent : map2.entrySet()){
            map.putIfAbsent(ent.getKey(), ent.getValue());
        }
        return map;
    }
    
    // Convert Char to Double
    // T1
    public HashMap<Double, Double> getMapDouble(HashMap<Double, Double> map){
        HashMap<Double, Double> m = new HashMap<>(Map.of(1.0, 1.0,
                2.0, 2.0, 3.0, 3.0));
        map = new HashMap<>(m);
        return m;
    }

    // T2
    public void clearMapDouble(HashMap<Double, Double> map, int size){
        if(map.isEmpty()) return;
        if(map.size() <= size) map.clear();
    }

    // T3
    public void containsThreeDouble(HashMap<Double, Double> map){
        if(map.containsKey(1.0) && map.containsKey(2.0) && map.containsKey(3.0)){
            map.put(4.0, 4.0);
        }
    }

    // T4
    public void containsDeleteDouble(HashMap<Double, Double> map, Double element){
        if(map.containsKey(element)){
            map.remove(element);
        }
    }

    public void containsDeleteListDouble(HashMap<Double, Double> map, List<Double> deleteList){
        for(Double element : deleteList){
            if(map.containsKey(element)){
                map.remove(element);
            }
        }
    }

    // T5
    public void tryAddSizeDouble(HashMap<Double, Double> map, int size, Double c){
        if(map.size() >= size) return;
        map.putIfAbsent(c, c);
    }

    public void tryAddListDouble(HashMap<Double, Double> map, List<Double> list){
        for(Double element : list){
            map.putIfAbsent(element, element);
        }
    }

    // T6
    public List<Double> getValuesListDoubleWithDefault(HashMap<Double, Double> map, List<Double> keys, Double default_){
        List<Double> values = new ArrayList<>();
        for(Double key : keys){
            Double value = map.getOrDefault(key, default_);
            values.add(value);
        }
        return values;
    }

    // T7
    public List<Optional<Double>> getValuesListDoubleWithNull(HashMap<Double, Double> map, List<Double> keys){
        List<Optional<Double>> values = new ArrayList<>();
        for(Double key : keys){
            Optional<Double> value = Optional.ofNullable(map.get(key));
            values.add(value);
        }
        return values;
    }

    // T8
    public HashMap<Double, Double> composeMapDouble(List<Double> keys, List<Double> values){
        HashMap<Double, Double> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<Double, String> composeMapDoubleString(List<Double> keys, List<String> values){
        HashMap<Double, String> map = new HashMap<>();
        if(keys.size() != values.size() || keys.size() == 0) return map;
        else{
            int size = keys.size();
            for(int i=0; i<size; ++i){
                map.put(keys.get(i), values.get(i));
            }
        }
        return map;
    }

    public HashMap<Double, Double> composeMapsDouble(HashMap<Double, Double> map1, HashMap<Double, Double> map2){
        HashMap<Double, Double> map = new HashMap<>();
        for(Map.Entry<Double, Double> ent : map1.entrySet()){
            map.put(ent.getKey(), ent.getValue());
        }

        for(Map.Entry<Double, Double> ent : map2.entrySet()){
            map.putIfAbsent(ent.getKey(), ent.getValue());
        }
        return map;
    }
}

File: Parallel-Csharp/src/Parallel/Containers/UnorderedMapParallel.cs
namespace Parallel_Csharp.Parallel.Containers;

public class UnorderedMapParallel {
    // T1
    public Dictionary<char, char> GetMapChar(Dictionary<char, char> map){
        Dictionary<char, char> m = new Dictionary<char, char>{{'a', 'a'},
            {'b', 'b'}, {'c', 'c'}};
        map = new Dictionary<char, char>(m);
        return m;
    }

    // T2
    public void ClearMapChar(Dictionary<char, char> map, int size){
        if(map.Count == 0) return;
        if(map.Count <= size) map.Clear();
    }

    // T3
    public void ContainsThreeChar(Dictionary<char, char> map){
        if(map.ContainsKey('a') && map.ContainsKey('b') && map.ContainsKey('c')){
            map['d'] = 'd';
        }
    }

    // T4
    public void ContainsDeleteChar(Dictionary<char, char> map, char element){
        if(map.ContainsKey(element)){
            map.Remove(element);
        }
    }

    public void ContainsDeleteListChar(Dictionary<char, char> map, List<char> deleteList){
        foreach(char element in deleteList){
            if(map.ContainsKey(element)){
                map.Remove(element);
            }
        }
    }

    // T5
    public void TryAddSizeChar(Dictionary<char, char> map, int size, char c){
        if(map.Count >= size) return;
        map.TryAdd(c, c);
    }

    public void TryAddListChar(Dictionary<char, char> map, List<char> list){
        foreach(char element in list){
            map.TryAdd(element, element);
        }
    }

    // T6
    public List<char> GetValuesListCharWithDefault(Dictionary<char, char> map, List<char> keys, char default_){
        List<char> values = new List<char>();
        foreach(char key in keys){
            char value = map.GetValueOrDefault(key, default_);
            values.Add(value);
        }
        return values;
    }

    // T7
    public List<char ?> GetValuesListCharWithNull(Dictionary<char, char> map, List<char> keys){
        List<char ?> values = new List<char ?>();
        foreach(char key in keys){
            char? value = map.ContainsKey(key)? map[key] : null;
            values.Add(value);
        }
        return values;
    }

    // T8
    public Dictionary<char, char> ComposeMapChar(List<char> keys, List<char> values){
        Dictionary<char, char> map = new Dictionary<char, char>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }

    public Dictionary<char, string> ComposeMapCharString(List<char> keys, List<string> values){
        Dictionary<char, string> map = new Dictionary<char, string>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }
    
    public Dictionary<char, char> ComposeMapsChar(Dictionary<char, char> map1, Dictionary<char, char> map2){
        Dictionary<char, char> map = new Dictionary<char, char>();
        foreach (KeyValuePair<char, char> pair in map1) {
            map[pair.Key] = pair.Value;
        }
        foreach (KeyValuePair<char, char> pair in map2) {
            map.TryAdd(pair.Key, pair.Value);
        }
        return map;
    }
    
    // Convert Char to String
    // T1
    public Dictionary<string, string> GetMapString(Dictionary<string, string> map){
        Dictionary<string, string> m = new Dictionary<string, string>{{"aa", "aa"},
            {"bb", "bb"}, {"cc", "cc"}};
        map = new Dictionary<string, string>(m);
        return m;
    }

    // T2
    public void ClearMapString(Dictionary<string, string> map, int size){
        if(map.Count == 0) return;
        if(map.Count <= size) map.Clear();
    }

    // T3
    public void ContainsThreeString(Dictionary<string, string> map){
        if(map.ContainsKey("aa") && map.ContainsKey("bb") && map.ContainsKey("cc")){
            map["dd"] = "dd";
        }
    }

    // T4
    public void ContainsDeleteString(Dictionary<string, string> map, string element){
        if(map.ContainsKey(element)){
            map.Remove(element);
        }
    }

    public void ContainsDeleteListString(Dictionary<string, string> map, List<string> deleteList){
        foreach(string element in deleteList){
            if(map.ContainsKey(element)){
                map.Remove(element);
            }
        }
    }

    // T5
    public void TryAddSizeString(Dictionary<string, string> map, int size, string c){
        if(map.Count >= size) return;
        map.TryAdd(c, c);
    }

    public void TryAddListString(Dictionary<string, string> map, List<string> list){
        foreach(string element in list){
            map.TryAdd(element, element);
        }
    }

    // T6
    public List<string> GetValuesListStringWithDefault(Dictionary<string, string> map, List<string> keys, string default_){
        List<string> values = new List<string>();
        foreach(string key in keys){
            string value = map.GetValueOrDefault(key, default_);
            values.Add(value);
        }
        return values;
    }

    // T7
    public List<string ?> GetValuesListStringWithNull(Dictionary<string, string> map, List<string> keys){
        List<string ?> values = new List<string ?>();
        foreach(string key in keys){
            string? value = map.GetValueOrDefault(key, null);
            values.Add(value);
        }
        return values;
    }

    // T8
    public Dictionary<string, string> ComposeMapString(List<string> keys, List<string> values){
        Dictionary<string, string> map = new Dictionary<string, string>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }

    public Dictionary<string, char> ComposeMapStringChar(List<string> keys, List<char> values){
        Dictionary<string, char> map = new Dictionary<string, char>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }
    
    public Dictionary<string, string> ComposeMapsString(Dictionary<string, string> map1, Dictionary<string, string> map2){
        Dictionary<string, string> map = new Dictionary<string, string>();
        foreach (KeyValuePair<string, string> pair in map1) {
            map[pair.Key] = pair.Value;
        }
        foreach (KeyValuePair<string, string> pair in map2) {
            map.TryAdd(pair.Key, pair.Value);
        }
        return map;
    }
    
    // Convert Char to Int
    // T1
    public Dictionary<int, int> GetMapInt(Dictionary<int, int> map){
        Dictionary<int, int> m = new Dictionary<int, int>{{1, 1},
            {2, 2}, {3, 3}};
        map = new Dictionary<int, int>(m);
        return m;
    }

    // T2
    public void ClearMapInt(Dictionary<int, int> map, int size){
        if(map.Count == 0) return;
        if(map.Count <= size) map.Clear();
    }

    // T3
    public void ContainsThreeInt(Dictionary<int, int> map){
        if(map.ContainsKey(1) && map.ContainsKey(2) && map.ContainsKey(3)){
            map[4] = 4;
        }
    }

    // T4
    public void ContainsDeleteInt(Dictionary<int, int> map, int element){
        if(map.ContainsKey(element)){
            map.Remove(element);
        }
    }

    public void ContainsDeleteListInt(Dictionary<int, int> map, List<int> deleteList){
        foreach(int element in deleteList){
            if(map.ContainsKey(element)){
                map.Remove(element);
            }
        }
    }

    // T5
    public void TryAddSizeInt(Dictionary<int, int> map, int size, int c){
        if(map.Count >= size) return;
        map.TryAdd(c, c);
    }

    public void TryAddListInt(Dictionary<int, int> map, List<int> list){
        foreach(int element in list){
            map.TryAdd(element, element);
        }
    }

    // T6
    public List<int> GetValuesListIntWithDefault(Dictionary<int, int> map, List<int> keys, int default_){
        List<int> values = new List<int>();
        foreach(int key in keys){
            int value = map.GetValueOrDefault(key, default_);
            values.Add(value);
        }
        return values;
    }

    // T7
    public List<int ?> GetValuesListIntWithNull(Dictionary<int, int> map, List<int> keys){
        List<int ?> values = new List<int ?>();
        foreach(int key in keys){
            int? value = map.ContainsKey(key)? map[key] : null;
            values.Add(value);
        }
        return values;
    }

    // T8
    public Dictionary<int, int> ComposeMapInt(List<int> keys, List<int> values){
        Dictionary<int, int> map = new Dictionary<int, int>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }

    public Dictionary<int, string> ComposeMapIntString(List<int> keys, List<string> values){
        Dictionary<int, string> map = new Dictionary<int, string>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }
    
    public Dictionary<int, int> ComposeMapsInt(Dictionary<int, int> map1, Dictionary<int, int> map2){
        Dictionary<int, int> map = new Dictionary<int, int>();
        foreach (KeyValuePair<int, int> pair in map1) {
            map[pair.Key] = pair.Value;
        }
        foreach (KeyValuePair<int, int> pair in map2) {
            map.TryAdd(pair.Key, pair.Value);
        }
        return map;
    }
    
    // Convert Char to Double
    // T1
    public Dictionary<double, double> GetMapDouble(Dictionary<double, double> map){
        Dictionary<double, double> m = new Dictionary<double, double>{{1.0, 1.0},
            {2.0, 2.0}, {3.0, 3.0}};
        map = new Dictionary<double, double>(m);
        return m;
    }

    // T2
    public void ClearMapDouble(Dictionary<double, double> map, int size){
        if(map.Count == 0) return;
        if(map.Count <= size) map.Clear();
    }

    // T3
    public void ContainsThreeDouble(Dictionary<double, double> map){
        if(map.ContainsKey(1.0) && map.ContainsKey(2.0) && map.ContainsKey(3.0)) {
            map[4.0] = 4.0;
        }
    }

    // T4
    public void ContainsDeleteDouble(Dictionary<double, double> map, double element){
        if(map.ContainsKey(element)){
            map.Remove(element);
        }
    }

    public void ContainsDeleteListDouble(Dictionary<double, double> map, List<double> deleteList){
        foreach(double element in deleteList){
            if(map.ContainsKey(element)){
                map.Remove(element);
            }
        }
    }

    // T5
    public void TryAddSizeDouble(Dictionary<double, double> map, int size, double c){
        if(map.Count >= size) return;
        map.TryAdd(c, c);
    }

    public void TryAddListDouble(Dictionary<double, double> map, List<double> list){
        foreach(double element in list){
            map.TryAdd(element, element);
        }
    }

    // T6
    public List<double> GetValuesListDoubleWithDefault(Dictionary<double, double> map, List<double> keys, double default_){
        List<double> values = new List<double>();
        foreach(double key in keys){
            double value = map.GetValueOrDefault(key, default_);
            values.Add(value);
        }
        return values;
    }

    // T7
    public List<double ?> GetValuesListDoubleWithNull(Dictionary<double, double> map, List<double> keys){
        List<double ?> values = new List<double ?>();
        foreach(double key in keys) {
            double? value = map.ContainsKey(key)? map[key] : null;
            values.Add(value);
        }
        return values;
    }

    // T8
    public Dictionary<double, double> ComposeMapDouble(List<double> keys, List<double> values){
        Dictionary<double, double> map = new Dictionary<double, double>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }

    public Dictionary<double, string> ComposeMapDoubleString(List<double> keys, List<string> values){
        Dictionary<double, string> map = new Dictionary<double, string>();
        if(keys.Count != values.Count || keys.Count == 0) return map;
        else{
            int size = keys.Count;
            for(int i=0; i<size; ++i){
                map[keys[i]] = values[i];
            }
        }
        return map;
    }
    
    public Dictionary<double, double> ComposeMapsDouble(Dictionary<double, double> map1, Dictionary<double, double> map2){
        Dictionary<double, double> map = new Dictionary<double, double>();
        foreach (KeyValuePair<double, double> pair in map1) {
            map[pair.Key] = pair.Value;
        }
        foreach (KeyValuePair<double, double> pair in map2) {
            map.TryAdd(pair.Key, pair.Value);
        }
        return map;
    }
}